### PR TITLE
RFC: Improve the way that locations are read

### PR DESCRIPTION
**Pull request by freben**
**BREAKING CHANGE**: If we merge this, the annotation `backstage.io/managed-by-location` will no longer point to a location ID, but to a location on the form `<type>:<target>`, e.g. `"github:https://github.com/spotify/backstage/blob/master/plugins/catalog-backend/fixtures/two_components.yaml"`. I believe that this is what we want to do long term.

The point here is that one "processor" may produce several different kinds of output. So now you are permitted to have for example a processor that sees a location and emits several other locations nearby (say, for a monorepo root, discovering all the yaml files under it). When reading a location, you may want to emit a bunch of entities, or raw data blobs that some other processor knows how to handle, or errors. This way you end up with a long "stream" of things to resolve, starting from a single location.

I thought this could be an interesting approach. It will give a greater amount of flexibility and code isolation than what we had, at the cost of the "pump" looking a bit weird.

### CODE REVIEW

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+      );
+      const current = { entity: item.entity, location: item.location };
+      await this.runAll(
+        'process entity',
+        processor =>
+          processor.processEntity?.(current.entity, current.location),
+        async emitted => {
+          if (emitted.type === 'entity') {
+            current.entity = emitted.entity;
+            current.location = emitted.location;
```
Comment (Rugvip) [inline plugins/catalog-backend/src/ingestion/LocationReaders.ts:145]: This'll override previously emitted entities right? Which I don't think is what we want

Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:145]: No it's right. We effectively want `emit(f1(f2(f3(f4(originalEntity))))`

**Comment by Rugvip** [comment]
So I'm not finding it easy to follow the logic, and I think a large part of that is that the logic for how to process items is heavily intertwined with the async control flow.

If we look at the processing bit and make it synchronous and dumb it down, it's something like this:

```js
function processItem(item, processors, outputs) {
  if (isProcessable(item)) {
    for (const process of processors) {
      for (const emitted of process(item)) {
        if (isOutput(emitted)) {
          outputs.push(emitted)
        } else {
          processItem(emitted, processors, outputs);
        }
      }
    }
  }
}

processItem(startItem, processors, [])
```

It's a lot of nesting back and forth and obviously becomes a lot more complex when it's all async.

I'm thinking of two things we could do to clean it up a bit. The first would be to split the async control flow out into a separate class, could be something like this:

```js
function process() {
  let entities = []

  const pipeline = makePipeline(processors)

  pipeline.submit(startItem)
  pipeline.on('output', entity => {
    entities.push(entity)
  })
  pipeline.on('emit', item => {
    pipeline.submit(item)
  })

  await pipeline.done()

  return entites
}
```

The other thing would be to distinguish between processor output that needs to be re-processed and output that can just move on to the next processor. Maybe something like this, where `yield`ing means reprocess by the full pipeline, but `return`ing just passes on to the next processor:

```js
function *processor(item) {
  if (!isInteresting(item)) {
    return item
  }

  if (isComposite(item)) {
    for async (const part of splitIntoParts(item)) {
      yield item
    }

    return undefined
  }

  item.myField = 3
  return item
}
```

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts (hunk 1 above)
Comment (Rugvip) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:145]: This is not a simple function composition though? What if `f2` emits 2 entities?

Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:145]: Yeah that's a point. Emitting more than one would be unexpected here, but unfortunately permitted by the api.

Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:145]: I could change the signature for processEntity to return `Promise<Entity>`

**Comment by freben** [comment]
Totally open to this being entangled :)

Would you think it more sensible if the processor methods were not generators, but instead returned `Promise<void>` and accepted a sink so that they could do `await emit(results.entity(...))` for example? 

That would unnest into a very straightforward back-and-forth dispatch with plain old promises.

**Comment by freben** [comment]
As a matter of fact - the `emit` callback could even be sync, effectively amounting to just a `push` behind the scenes - so that the driver just gathers up a list of output in the background and then works through them one by one, in any way it sees fit.

**Comment by Rugvip** [comment]
Yeah my last example could easily just use an emit instead of yield. Still think it's valuable to differ between new results and transformations though

**Comment by freben** [comment]
I still think that there's only one thing that ever does NOT get put back at the start of the processing chain, and that's entities that have multiple processors that may want to mutate/inspect them before they are put in the result set.

Therefore I propose that the signature for processEvent be different. It should return `Promise<Event>` instead, and be pumped explicitly like that. They may optionally ALSO be permitted to emit other things with a callback like the others, and that gets reconsidered from the top as usual, but by contract they have to return the same entity or a modified version of it.

**Comment by freben** [comment]
(an intermediate version of the code looked somewhat like that but then I admittedly over-generalized)

**Comment by Rugvip** [comment]
Sounds good, as long as they can also emit an error instead of an entity. Might just be done with throw?

**Comment by freben** [comment]
I mean... At that point yes maybe.

But that makes one wonder - should ALL processor variants just return a results container instead :)

**Comment by Rugvip** [comment]
yeah that'd work too

**Comment by freben** [comment]
@Rugvip straightened out things - much more controlled complexity now

File: plugins/catalog-backend/src/ingestion/processors/types.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+  /**
+   * Handles an emitted error.
+   *
+   * @param error The error
+   * @param location The location where the error occurred
+   * @param emit A sink for items resulting from this handilng
+   * @returns Nothing
+   */
+  handleError?(
```
Comment (Rugvip) [inline plugins/catalog-backend/src/ingestion/processors/types.ts:70]: This isn't used atm right? What's a use-case you had in mind?

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+      for (const item of items) {
+        if (item.type === 'location') {
+          await this.handleLocation(item, sink);
+        } else if (item.type === 'data') {
+          await this.handleData(item, sink);
+        } else if (item.type === 'entity') {
+          await this.handleEntity(item, sink, output);
+        } else if (item.type === 'error') {
+          await this.handleError(item, sink, output);
```
Comment (Rugvip) [inline plugins/catalog-backend/src/ingestion/LocationReaders.ts:87]: I don't quite get the purpose of the split into these 4 handle methods tbh. As opposed to just letting the processor inspect the type and act accordingly.

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    this.processors = processors;
+  }
+
+  async read(location: LocationSpec): Promise<ReadLocationResult> {
+    const output: ReadLocationResult = { entities: [], errors: [] };
+    let items: LocationProcessorResult[] = [result.location(location, false)];
+
+    for (let depth = 0; depth < MAX_DEPTH; ++depth) {
+      const newItems: LocationProcessorResult[] = [];
+      const sink: LocationProcessorSink = i => newItems.push(i);
```
Comment (Rugvip) [inline plugins/catalog-backend/src/ingestion/LocationReaders.ts:77]: Could call it `emit` here too, and possibly have the type called `LocationProcessorEmit`? Kinda weird that it switches name all of a sudden without really switching context much.

**Review by Rugvip — COMMENTED** [review]
Yup, easier to follow :grin:

File: plugins/catalog-backend/src/ingestion/processors/types.ts (hunk 2 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/processors/types.ts:70]: Right. Things like metrics was what I had in mind.

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts (hunk 3 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:87]: The "driver" contains the behaviors like "stop after first match" and similar. The processor is not aware of that coordination logic.

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts (hunk 4 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:77]: Agreed, that was sloppy naming.

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts (hunk 3 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:87]: Oh reading this again, you mean that there would be an extra `if` condition inside a single universal method in the processor? Well they still have different signatures, right

Comment (Rugvip) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:87]: Yeah I'm questioning why they need to have different signatures, instead of just a `handleItem` that returns zero or one item and optionally emits additional things

Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:87]: What does returning zero items mean? What does returning one item mean? Is there a difference between returning one Buffer, and returning one Entity? Or returning one Entity and emitting one error? Etc. I am not sure how you'd stitch that up, man, sounds way more confusing

Comment (Rugvip) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts:87]: Returning one item would mean the next processor in the chain picks it up. Returning zero means there's no further processing of it. Emitting items always sends them to the beginning of the chain. Throwing an error would then mean there's an error, and emitting errors I don't think would be a thing.

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts (hunk 1 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts]: changed this to do explicit chaining in handleEntity

File: plugins/catalog-backend/src/ingestion/LocationReaders.ts (hunk 3 above)
Comment (freben) [reply plugins/catalog-backend/src/ingestion/LocationReaders.ts]: attempted this in #1179 but actually am not fond of it

**Review by Rugvip — APPROVED** [review]
